### PR TITLE
chore: Faster LRU implementation, and more aggressive caching of IPFS access

### PR DESCRIPTION
**Pull request by ukstv**
Note https://github.com/ceramicnetwork/js-ceramic/pull/2903/files#diff-3dc086303a234c0aa9cb5b7b4b675827819e66906135448b777b221da01e0d60R55

`applyCommit` reuses IPLD records it got.

### CODE REVIEW

File: packages/core/src/ancillary/ipld-records-cache.ts [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+   */
+  get(cid: CID): CacheEntry | undefined {
+    return this.cache.get(cid.toString())
+  }
+
+  /**
+   * IPFS allows you to get an IPLD record by query like `CID/some/path`. We do not have traversal mechanism,
+   * so we just store an entry by query. See also `getR`.
+   */
+  setR(key: string, entry: CacheEntry): void {
```
Comment (dbcfd) [inline packages/core/src/ancillary/ipld-records-cache.ts:70]: Not a fan of this name, maybe `setIPLDRecord`?

File: packages/core/src/ancillary/ipld-records-cache.ts [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+    this.cache.set(key, entry)
+  }
+
+  /**
+   * IPFS allows you to get an IPLD record by query like `CID/some/path`. We do not have traversal mechanism,
+   * so here we do our best to return an entry by query. See also `setR`.
+   */
+  getR(key: string): CacheEntry | undefined {
```
Comment (dbcfd) [inline packages/core/src/ancillary/ipld-records-cache.ts:78]: Same as above

File: packages/core/src/state-management/state-cache.ts [hunk 3]
```diff
@@ -1,23 +1,23 @@
...
    */
   constructor(limit: number, private readonly onEvicted?: (item: T) => void) {
-    this.volatile = new lru.LRUMap(limit)
-    this.volatile.shift = function () {
-      const entry = lru.LRUMap.prototype.shift.call(this)
-      onEvicted?.(entry[1])
-      return entry
+    this.volatile = new LRUCache(limit)
+    if (onEvicted) {
+      this.volatile.on('evicted', (_key, value) => {
```
Comment (dbcfd) [inline packages/core/src/state-management/state-cache.ts:18]: Can we add eviction metrics?

File: packages/core/src/store/pin-store-factory.ts [hunk 4]
```diff
@@ -39,6 +41,8 @@ export class PinStoreFactory {
     const ipfs = this.ipfs
     const pinning = PinningAggregation.build(ipfs, this.pinningEndpoints, this.pinningBackends)
     const retrieve = async (cid: CID): Promise<any> => {
+      const fromCache = this.ipldRecordsCache.get(cid)
+      if (fromCache) return fromCache.record
```
Comment (dbcfd) [inline packages/core/src/store/pin-store-factory.ts:45]: Can we add metrics here?

File: packages/core/src/dispatcher.ts [hunk 5]
```diff
@@ -167,16 +167,21 @@ export class Dispatcher {
...
   }
 
   async getIpfsBlock(cid: CID): Promise<Uint8Array> {
-    return await this._shutdownSignal.abortable((signal) => {
-      // @ts-ignore
-      return this._ipfs.block.get(cid, { signal, offline: !this.enableSync })
-    })
+    const found = this.ipldCache.get(cid)
+    if (found) {
+      return found.block
```
Comment (dbcfd) [inline packages/core/src/dispatcher.ts:178]: Can we add metrics here (think this is the right spot)?

**Review by dbcfd — CHANGES_REQUESTED** [review]

File: packages/core/src/store/pin-store-factory.ts (hunk 4 above)
Comment (ukstv) [reply packages/core/src/store/pin-store-factory.ts]: Uhums! Added

File: packages/core/src/dispatcher.ts (hunk 5 above)
Comment (ukstv) [reply packages/core/src/dispatcher.ts:179]: Added.

**Review by dbcfd — APPROVED** [review]

File: packages/core/src/dispatcher.ts [hunk 6]
```diff
@@ -306,8 +327,12 @@ export class Dispatcher {
 
     // Lookup CID in cache before looking it up IPFS
     const resolutionPath = `${asCid}${path || ''}`
-    const cachedDagNode = this.dagNodeCache.get(resolutionPath)
-    if (cachedDagNode) return cloneDeep(cachedDagNode)
+    const fromCache = this.ipldCache.getUnresolved(resolutionPath)
+    if (fromCache) {
+      Metrics.count(IPFS_CACHE_HIT, 1)
+      return cloneDeep(fromCache.record)
```
Comment (stbrody) [inline packages/core/src/dispatcher.ts:333]: do you think we really need to clone this here?  How likely is it that any callers are modifying the record?

File: packages/core/src/store/ipld-records-cache.ts [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+   */
+  get(cid: CID): CacheEntry | undefined {
+    return this.cache.get(cid.toString())
+  }
+
+  /**
+   * IPFS allows you to get an IPLD record by a query like `CID/some/path`. We do not have resolution mechanism that would translate `cid/path` query into a CID,
+   * so we just store an entry by query. See also `getUnresolved`.
+   */
+  setUnresolved(query: string, entry: CacheEntry): void {
```
Comment (stbrody) [inline packages/core/src/store/ipld-records-cache.ts:73]: nitpick: maybe `setWithResolutionPath` is clearer?

File: packages/core/src/store/pin-store-factory.ts [hunk 8]
```diff
@@ -39,6 +43,12 @@ export class PinStoreFactory {
     const ipfs = this.ipfs
     const pinning = PinningAggregation.build(ipfs, this.pinningEndpoints, this.pinningBackends)
     const retrieve = async (cid: CID): Promise<any> => {
+      const fromCache = this.ipldRecordsCache.get(cid)
+      if (fromCache) {
+        Metrics.count(IPFS_CACHE_HIT, 1)
+        return fromCache.record
+      }
+      Metrics.count(IPFS_CACHE_MISS, 1)
       const blob = await ipfs.dag.get(cid, { timeout: IPFS_GET_TIMEOUT })
```
Comment (stbrody) [inline packages/core/src/store/pin-store-factory.ts:52]: Should we add the entry into the cache here?

File: packages/core/src/store/pin-store-factory.ts [hunk 9]
```diff
@@ -39,6 +43,12 @@ export class PinStoreFactory {
     const ipfs = this.ipfs
     const pinning = PinningAggregation.build(ipfs, this.pinningEndpoints, this.pinningBackends)
     const retrieve = async (cid: CID): Promise<any> => {
+      const fromCache = this.ipldRecordsCache.get(cid)
```
Comment (stbrody) [inline packages/core/src/store/pin-store-factory.ts:46]: Ideally at some point I'd like to pull out the parts of the Dispatcher that just manage direct puts and gets into IPFS (and add the caching, timeout, and retry logic) into a separate IpfsService API and then use that here directly so we can pick have unified access to the caching logic and also pick up the other extras that the Dispatcher adds over raw IPFS access.  But that is certainly out of scope for this PR.

**Review by stbrody — APPROVED** [review]
nice

File: packages/core/src/store/ipld-records-cache.ts (hunk 7 above)
Comment (ukstv) [reply packages/core/src/store/ipld-records-cache.ts:73]: Oh! Finally a good name for this method! Thank you!

File: packages/core/src/store/pin-store-factory.ts (hunk 9 above)
Comment (ukstv) [reply packages/core/src/store/pin-store-factory.ts:46]: Someday, yeah. That’s on my list too.

File: packages/core/src/dispatcher.ts (hunk 6 above)
Comment (ukstv) [reply packages/core/src/dispatcher.ts:333]: Haha, I had the same question to myself. I thought you would be against removing it :)

Comment (stbrody) [reply packages/core/src/dispatcher.ts:333]: Ideally the type system would let us specify the return value as immutable, but alas

Comment (dbcfd) [reply packages/core/src/dispatcher.ts:333]: Just sprinkle some rust on it 😆

**Comment by ukstv** [comment]
Nope, cloneDeep is here to stay. Absence of cloneDeep breaks the tests. IDK what's wrong, but it's another pain point better addressed separately.

**Pull request merged**